### DISCUSSION TITLE
Callout boxes for ipynb output

### DISCUSSION THREAD

**Discussion by linogaliana** [quarto-dev/quarto-cli#1166] [Feature Requests]
Hi,

`callouts` are a really nice feature. I have seen in the doc that currently Jupyter Notebooks are not supported (https://quarto.org/docs/authoring/callouts.html). I totally understand this is not a priority and might be tricky to implement. However, in the future, that would be really cool to also propose boxes for this output format.

- As far as I understood discussions in https://github.com/jupyter/notebook/issues/1292 it seems that there exists some features in Jupyter Notebook with some raw HTML `<div>`. Some examples are given [here](https://github.com/jupyter/notebook/issues/1292#issuecomment-570548806)
- However, all boxes provided by `quarto` are not supported
- Boxes are not as aesthetic as those provided by `quarto` (or jupyterbook). This is of course not a priority (functional boxes would first be great) but this would be cool to a nice box in notebook. I saw that it is possible to embed some styling element directly in the `div` ([see this stackoverflow thread](https://stackoverflow.com/questions/62562105/syntax-highlighting-for-code-blocks-within-alert-box-in-jupyter-for-both-noteboo)) but maybe there's a better way to start
- @choldgraf maybe has an opinion on the best way to proceed

Once again, I understand that tweaking some notebook features is not a priority. But helping python users generate nice notebooks from quarto could be appreciated

**Comment by choldgraf** [comment]
Hey @linogaliana thanks for pinging me.

I think that your best bet would be to develop [a JupyterLab extension](https://jupyterlab.readthedocs.io/en/stable/user/extensions.html) that handled the parsing / rendering of quarto callouts within a JupyterLab interface. If there's a javascript/typescript parser for it, it shouldn't be too tough.

As one example, the community has been building out a MyST renderer for JupyterLab, which piggy-backs on a generic MyST parser/renderer in JavaScript (MyST is the cross-platform / extensible flavor of markdown that is also used in the Jupyter Book project). I'll link below in case it's helpful for inspiration:

- Lab extension to render MyST: https://github.com/executablebooks/jupyterlab-myst
- MyST JS parser/renderer: https://github.com/executablebooks/mystjs

**Comment by jjallaire** [comment]
I agree that having a JupyterLab extension for this would be a huge step forward! That would make authoring of callouts in JupyterLab much nicer (as they currently are for MyST).

It doesn't solve this particular problem though, as this scenario covers the generation of ipynb notebooks from quarto documents (which could be edited using vi, emacs, vscode, rstudio, jupyterlab, etc.). Here the actual published output is ipynb (rather than the editing input being ipynb -- it certainly might be but it could also be a .qmd text file) so we need the quarto engine to be able to render  into a raw html output in Jupyter.

**Reply by choldgraf** [reply]
ahhh gotcha, I mis-understood the original question, sorry for the noise!

**Comment by mcanouil** [comment]
Closing in favour of https://github.com/quarto-dev/quarto-cli/issues/1167.
